server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistoryService.java

```diff
@@ -0,0 +1,122 @@
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.LifecycleListener;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * This service provides access to this node's view of the master history, as well as access to other nodes' view of master stability.
+ */
+public class MasterHistoryService {
+    private final TransportService transportService;
+    private final MasterHistory localMasterHistory;
+
+    public MasterHistoryService(
+        TransportService transportService,
+        Coordinator coordinator,
+        ThreadPool threadPool,
+        ClusterService clusterService
+    ) {
+        this.transportService = transportService;
+        this.localMasterHistory = new MasterHistory(threadPool, clusterService);
+        // Set the initial state for the local history once it is available:
+        coordinator.addLifecycleListener(new LifecycleListener() {
+            @Override
+            public void afterStart() {
+                localMasterHistory.clusterChanged(
+                    new ClusterChangedEvent(MasterHistoryService.class.getName(), clusterService.state(), clusterService.state())
+                );
+            }
+        });
+    }
+
+    /**
+     * This returns the MasterHistory as seen from this node. The returned MasterHistory will be automatically updated whenever the
+     * ClusterState on this node is updated with new information about the master.
+     * @return The MasterHistory from this node's point of view. This MasterHistory object will be updated whenever the ClusterState changes
+     */
+    public MasterHistory getLocalMasterHistory() {
+        return localMasterHistory;
+    }
+
+    /**
+     * This method returns a static view of the MasterHistory on the node given. This MasterHistory is static in that it will not be
+     * updated even if the ClusterState is updated on this node or the remote node.
+     * @param node The node whose MasterHistory we want
+     * @return The MasterHistory from the remote node's point of view. This MasterHistory object will not be updated with future changes
+     * @throws ExecutionException If an exception occurs while trying to fetch the remote node's MasterHistory
+     * @throws InterruptedException If the thread is interrupted
+     */
+    public List<DiscoveryNode> getRemoteMasterHistory(DiscoveryNode node) throws ExecutionException, InterruptedException {
+        MasterHistoryAction.Request getMasterHistoryRequest = new MasterHistoryAction.Request();
+        final MasterHistoryAction.Response[] responseArray = new MasterHistoryAction.Response[1];
+        final Exception[] exceptionArray = new Exception[1];
+        final Object lock = new Object();
+        transportService.connectToNode(node, new ActionListener<>() {
```

Comment:
I'm guessing there's a better way to do this.

Reply:
It'd be better not to block a thread here, which involves making this method async:

```public void getRemoteMasterHistory(DiscoveryNode node, ActionListener<List<DiscoveryNode>> listener)```

That change needs to propagate all the way up the call chain too. I'd also recommend using the variant of `transportService.sendRequest` which takes a `TransportRequestOptions` so you can set a timeout.

Reply:
> That change needs to propagate all the way up the call chain too

Oh, but we're calling this from `StableMasterHealthIndicatorService#compute` which is synchronous, and that is deliberate because `GetHealthAction` itself is supposed to be lightweight. We can't have this API wait for a remote node to respond. We'll have to work out the master history & stability in the background and just expose the latest status somewhere that the health indicator can read.

Reply:
In this case is that not OK? I assume it's pretty rare that you get here, and if you do get here then things are in pretty bad shape. So is waiting a few seconds for the right answer such a bad thing? Is this type of remote request likely to take more than a couple of seconds? Basically I'm just wondering if doing this asynchronously is worth the added complexity?

Reply:
Alright as we discussed, I changed this to asynchronously fetch the remote master history in the MasterHistoryService's clusterChanged method when it looks like we might need it (the master has gone null and it has repeatedly gone null in the last 30 minutes). A big downside is that it couples the implementation of the MasterHistoryService to what the StableMasterHistoryIndicatorService needs -- that is, the remote history just won't be there unless the criteria that StableMasterHistoryIndicatorService cares about (master has gone null more than 3 times in 30 minutes) are met. I probably need to put some more thought into that. Maybe let the StableMasterHistoryIndicatorService decide that it will need the remote history, and request that MasterHistoryService fetch it.

Reply:
OK, I moved the logic for deciding whether we needed a remote master's history into StableMasterHistoryIndicatorService, and left the logic for actually fetching it in MasterHistoryService. So now there's one method in MasterHistoryService you hit to say "we might need this node's master history in the future", and another method in MasterHistoryService to say "ok give me what master history you have for this node".